servers/rendering/shader_preprocessor.cpp

```diff
@@ -0,0 +1,1074 @@
+					// This ensures we don't skip the other args of this macro in the string.
+					arg_index_start = arg_index + args[i].length() + 1;
+				}
+			}
+
+			result = result.substr(0, index) + " " + body + " " + result.substr(args_end + 1, result.length());
+		} else {
+			result = result.substr(0, index) + body + result.substr(index + key.length(), result.length() - (index + key.length()));
+			// Manually reset index_start to where the body value of the define finishes.
+			// This ensures we don't skip another instance of this macro in the string.
+			index_start = index + body.length() + 1;
+			break;
+		}
+	}
+	r_expanded = result;
+	return OK;
+}
+
+bool ShaderPreprocessor::find_match(const String &p_string, const String &p_value, int &r_index, int &r_index_start) {
+	// Looks for value in string and then determines if the boundaries
+	// are non-word characters. This method semi-emulates \b in regex.
+	r_index = p_string.find(p_value, r_index_start);
+	while (r_index > -1) {
+		if (r_index > 0) {
+			if (is_char_word(p_string[r_index - 1])) {
+				r_index_start = r_index + 1;
+				r_index = p_string.find(p_value, r_index_start);
+				continue;
+			}
+		}
+
+		if (r_index + p_value.length() < p_string.length()) {
+			if (is_char_word(p_string[r_index + p_value.length()])) {
+				r_index_start = r_index + p_value.length() + 1;
+				r_index = p_string.find(p_value, r_index_start);
+				continue;
+			}
+		}
+
+		// Return and shift index start automatically for next call.
+		r_index_start = r_index + p_value.length() + 1;
+		return true;
+	}
+
+	return false;
+}
+
+String ShaderPreprocessor::next_directive(Tokenizer *p_tokenizer, const Vector<String> &p_directives) {
+	const int line = p_tokenizer->get_line();
+	int nesting = 0;
+
+	while (true) {
+		p_tokenizer->advance('#');
+
+		String id = p_tokenizer->peek_identifier();
+		if (id.is_empty()) {
+			break;
+		}
+
+		if (nesting == 0) {
+			for (int i = 0; i < p_directives.size(); i++) {
+				if (p_directives[i] == id) {
+					p_tokenizer->backtrack('#');
+					return id;
+				}
+			}
+		}
+
+		if (id == "ifdef" || id == "ifndef" || id == "if") {
+			nesting++;
+		} else if (id == "endif") {
+			nesting--;
+		}
+	}
+
+	set_error(RTR("Can't find matching branch directive."), line);
+	return "";
+}
+
+void ShaderPreprocessor::add_to_output(const String &p_str) {
+	for (int i = 0; i < p_str.length(); i++) {
+		output.push_back(p_str[i]);
+	}
+}
+
+void ShaderPreprocessor::set_error(const String &p_error, int p_line) {
+	if (state->error.is_empty()) {
+		state->error = p_error;
+		FilePosition fp;
+		fp.line = p_line + 1;
+		state->include_positions.push_back(fp);
+	}
+}
+
+bool ShaderPreprocessor::check_directive_before_type(Tokenizer *p_tokenizer, const String &p_directive) {
+	if (p_tokenizer->get_index() < state->shader_type_pos) {
+		set_error(vformat(RTR("`#%s` may not be defined before `shader_type`."), p_directive), p_tokenizer->get_line());
+		return false;
+	}
+	return true;
+}
```

Comment:
This function does not needed now, since I've removed the handling of `shader_type_pos` (which prevent declaration of directives before shader type).